cache/cache_test.go

```diff
@@ -398,35 +399,84 @@ func TestValidateWatchRange(t *testing.T) {
 	}
 }
 
-type mockWatcher struct {
-	responses  chan clientv3.WatchResponse
-	registered chan struct{}
-}
+func TestCacheCompactionResync(t *testing.T) {
+	firstSnapshot := &clientv3.GetResponse{
+		Header: &pb.ResponseHeader{Revision: 5},
+		Kvs:    []*mvccpb.KeyValue{{Key: []byte("foo"), ModRevision: 5}},
+	}
+	secondSnapshot := &clientv3.GetResponse{
+		Header: &pb.ResponseHeader{Revision: 20},
+		Kvs:    []*mvccpb.KeyValue{{Key: []byte("foo"), ModRevision: 20}},
+	}
+	fakeClient := &clientv3.Client{
+		Watcher: newMockWatcher(16),
+		KV:      newKVStub(firstSnapshot, secondSnapshot),
+	}
+	cache, err := New(fakeClient, "")
+	if err != nil {
+		t.Fatalf("New cache: %v", err)
+	}
+	defer cache.Close()
+	mw := fakeClient.Watcher.(*mockWatcher)
 
-type kvStub struct{}
+	t.Log("Phase 1: initial getWatch bootstrap")
+	mw.triggerCreatedNotify()
+	<-mw.registered
+	if err = cache.WaitReady(t.Context()); err != nil {
+		t.Fatalf("initial WaitReady: %v", err)
+	}
 
-func (kvStub) Get(ctx context.Context, key string, _ ...clientv3.OpOption) (*clientv3.GetResponse, error) {
-	return &clientv3.GetResponse{Header: &pb.ResponseHeader{Revision: 0}}, nil
-}
+	t.Log("Phase 2: simulate compaction")
+	mw.errorCompacted(10)
+	waitForResetReady(t, cache, 100*time.Millisecond)
+	if gotRev := cache.store.LatestRev(); gotRev != 0 {
```

Comment:
Can you test Get request instead of inspecting internals? We should expect that Get should timeout waiting for ready.